### DISCUSSION TITLE
Next.js OpenTelemetry auto-instrumentation does not instrument fetch API correctly

### DISCUSSION THREAD

**Discussion by matteopaoli** [vercel/next.js#57932] [Help]
### Summary

Auto-instrumentation of Fetch API doesn't come with W3C propagation for trace context. I also tried different third-party instrumentation libraries and none of them seem to do that, I tried with:

- @opentelemetry/auto-instrumentations-node
- opentelemetry-instrumentation-undici

Since Next.js provides its own implementation of Fetch API on the server I think its instrumentation library should inject the propagation headers when the feature is on. I ended up with writing my own logic for that.

**Comment by icyJoseph** [comment]
> Since Next.js provides its own implementation of Fetch API on the server 

It doesn't anymore right? or rather, if you are in Node 18+ then `fetch` is available natively on the Node.js environment. 

Have you checked if the problem is not within undici? https://therootcompany.com/blog/node-native-fetch-undici/

**Reply by matteopaoli** [reply]
i suppose it still does because the fetch function is enhanced with additional parameters

https://nextjs.org/docs/app/api-reference/functions/fetch#optionsnextrevalidate

**Reply by icyJoseph** [reply]
That's monkey patching, but it doesn't mean it has its own implementation of the fetch API. Next.js when possible uses the environments fetch API.

**Comment by fbnptrs** [comment]
We do have the same problem, we've come up with a workaround that works, but would be nice to have it fixed in Next itself, see https://github.com/vercel/next.js/issues/51233#issuecomment-1895446474
